PAM: add pam_response_filter option

sumit-bose opened the pull request:
Currently the main use-case for this new option is to not set the
KRB5CCNAME environment varible for services like 'sudo-i'.

Resolves https://fedorahosted.org/sssd/ticket/2296

jhrozek commented:
retest this please

jhrozek commented:
Thank you for the patch, I started the review and so far I would like to ask for compilation warnings to be fixed:
```
/home/remote/jhrozek/devel/sssd/src/tests/cmocka/test_pam_srv.c: In function ‘test_filter_response’:
/home/remote/jhrozek/devel/sssd/src/tests/cmocka/test_pam_srv.c:1783:17: warning: assignment discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
     pd->service = "MyService";
                 ^
/home/remote/jhrozek/devel/sssd/src/tests/cmocka/test_pam_srv.c:1793:5: warning: ‘memset’ used with constant zero length parameter; this could be due to transposed parameters [-Wmemset-transposed-args]
     memset(offline_auth_data, sizeof(offline_auth_data), 0);
     ^~~~~~
  CC       src/responder/pam/pam_srv_tests-pamsrv_dp.o
In file included from /home/remote/jhrozek/devel/sssd/src/util/util.h:54:0,
                 from /home/remote/jhrozek/devel/sssd/src/responder/pam/pamsrv_cmd.c:24:
/home/remote/jhrozek/devel/sssd/src/responder/pam/pamsrv_cmd.c: In function ‘filter_responses_env’:
/home/remote/jhrozek/devel/sssd/src/responder/pam/pamsrv_cmd.c:522:15: warning: field precision specifier ‘.*’ expects argument of type ‘int’, but argument 6 has type ‘size_t {aka long unsigned int}’ [-Wformat=]
               "Found PAM ENV filter for variable [%.*s] and service [%s].\n",
               ^
/home/remote/jhrozek/devel/sssd/src/util/debug.h:108:22: note: in definition of macro ‘DEBUG’
                      format, ##__VA_ARGS__); \
                      ^~~~~~
```

jhrozek commented:
Additionally, it looks like the reason CI failed is a valgrind issue in one of the tests:
```
[ RUN      ] test_filter_response
==19061== Conditional jump or move depends on uninitialised value(s)
==19061==    at 0x41BB2B: filter_responses (pamsrv_cmd.c:604)
==19061==    by 0x418B41: test_filter_response (test_pam_srv.c:1804)
==19061==    by 0x4E3C542: cmocka_run_one_test_or_fixture (cmocka.c:2305)
==19061==    by 0x4E3CC7A: cmocka_run_one_tests (cmocka.c:2413)
==19061==    by 0x4E3CC7A: _cmocka_run_group_tests (cmocka.c:2518)
==19061==    by 0x419B73: main (test_pam_srv.c:2082)
==19061== 
==19061== Conditional jump or move depends on uninitialised value(s)
==19061==    at 0x41BB3C: filter_responses (pamsrv_cmd.c:605)
==19061==    by 0x418B41: test_filter_response (test_pam_srv.c:1804)
==19061==    by 0x4E3C542: cmocka_run_one_test_or_fixture (cmocka.c:2305)
==19061==    by 0x4E3CC7A: cmocka_run_one_tests (cmocka.c:2413)
==19061==    by 0x4E3CC7A: _cmocka_run_group_tests (cmocka.c:2518)
==19061==    by 0x419B73: main (test_pam_srv.c:2082)
==19061== 
[       OK ] test_filter_response

```

jhrozek commented:
btw except from the warnings and the valgrind issue, the code seems to work fine. I tested with sudo -i and then kinit, before the patch:
```
-sh-4.3$ sudo -iu admin
[sudo] password for tuser: 
-bash-4.3$ echo $KRB5CCNAME
FILE:/tmp/krb5cc_1475800004_Pvl1cf
-bash-4.3$ kinit admin
Password for admin@IPA.TEST: 
kinit: Failed to store credentials: Internal credentials cache error (filename: /tmp/krb5cc_1475800004_Pvl1cf) while getting initial credentials
```

After the patch:
```
sh-4.3$ sudo -iu admin
[sudo] password for tuser: 
-bash-4.3$ echo $KRB5CCNAME

-bash-4.3$ kinit admin
Password for admin@IPA.TEST: 
-bash-4.3$ logout
```

jhrozek commented:
Additionally, do you think we should open a ticket to track potentially adding sudo -i to the list by default?

sumit-bose commented:
Thank you for the review, I fixed the warnings and updated my VM for F25 so that I can see the warnings as well :-) (expect the -Wformat which is only shown on 64bits). The valgrind warning was a side effect of the wrong usage of memset().

jhrozek commented:
Thank you, the patches still work and Coverity didn't find anything either.

ACK

jhrozek commented:
CI: http://sssd-ci.duckdns.org/logs/job/56/16/summary.html

jhrozek commented:
* master:
 * ce43f710c9638fbbeae077559cd7514370a10c0c
 * c8fe1d922b254aa92e74f428135ada3c8bde87a1

jhrozek commented:
* sssd-1-14:
 * 74711db46029415cc9590bb0e3f9cc662dac1d0c
 * 0157678081e299660105c753f2d2ac2081960bca

jhrozek commented:
@sumit-bose the patches do not apply cleanly on top of the sssd-1-13 branch, it looks like there is a conflict in the INI rules code. Could you please push patches for sssd-1-13 as well to some branch?

sumit-bose commented:
I guess this is because the file does not exists in 1.13. Please find patches on top of sssd-1-13 at https://github.com/sumit-bose/sssd/tree/filter_KRB5CCNAME-1-13

jhrozek commented:
thank you, pushed to sssd-1-13:
sssd-1-13:
 * 5cbea81f6019d4c6acc94b8cec0b955d1767430b
 * 27e38ce07c3f2bb9682e2219ac2ac78e855d43b5
